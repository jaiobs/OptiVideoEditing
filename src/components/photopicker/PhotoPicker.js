import {
  ActivityIndicator,
  Dimensions,
  FlatList,
  Image,
  NativeModules,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from "react-native";
import { Colors, Images, Strings } from "../../res";
import {
  DataProvider,
  LayoutProvider,
  RecyclerListView,
} from 'recyclerlistview';
import React, { Component } from "react";
import { checkAndRequestPermission } from '../../utils/PermissionUtils';
import { PERMISSIONS, openSettings } from 'react-native-permissions';

const INITIAL_LOAD_IMAGES_AHEAD = 18;
const LOAD_IMAGES_AHEAD = 6;
const screenWidth = Dimensions.get('window').width;


export class ImageCell extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const { disabled, selected, onSelect, imageUrl, selectedIndex } = this.props;

    return (
      <TouchableOpacity
        disabled={disabled}
        onPress={() => {
          onSelect();
        }}
        style={{
          width: "100%",//(this.state.width - 8) / this.props.rowItems,
          height: "100%", //this.state.width / this.props.rowItems,
          borderColor: Colors.primaryAccent,
          borderWidth: selected ? 3 : 0,
          margin: 2,
          padding: 1
        }}
      >
        <Image
          source={{ uri: imageUrl }}
          style={{ width: "100%", height: "100%" }}
        />

    {selectedIndex == -1 ? null : (
      <View style={{ position: "absolute", top: 5, right: 5, width:20, height:20, borderRadius:10, backgroundColor: Colors.primaryAccent, justifyContent:'center', alignItems:'center', alignContent:'center' }}>
        <Text style={{ color: "white", fontSize: 12 }}>
          {selectedIndex + 1}
        </Text>
      </View>
    )}
      </TouchableOpacity>
    );
  }
}

export default class PhotoPicker extends Component {
  CameraRoll = NativeModules.RNCCameraRoll;

  constructor(props) {
    super(props);
    this._layoutProvider = new LayoutProvider(
      this.getLayoutTypeForIndex,
      this.setLayoutForType
    );
    this._dataProvider = new DataProvider(this.rowHasChanged, this.getStableId);
    this.state = { 
      finishedLoading: false,
      rows: [],
      selectedFiles: [],
      selectedIndexes: [],
      lastCursor: "",
      hasNextpage: true,
      width: props.width,
      dataProvider: this._dataProvider.cloneWithRows([]),
      permissionType:'',
      permissionFlag:false
    };
  }


   permissionCheck = async () => {
    try {
      const hasPhotosAccess = await checkAndRequestPermission(
        PERMISSIONS.IOS.PHOTO_LIBRARY
      );
      this.setState({
        permissionFlag: hasPhotosAccess,
      });
      return hasPhotosAccess;
    } catch (err) {
      console.log('permission error -> ', err);
      return false;
    }
  };

  componentDidMount() {
    this.permissionCheck();
    const params = {
      first: INITIAL_LOAD_IMAGES_AHEAD,
      assetType: "Photos",
      groupTypes: 'All'
    };
    
    this.getGalleryDetails(params);
  }

  rowHasChanged = (r1,r2) => {
    return r1.isSelected !== r2.isSelected || r1.node.image.filename !== r2.node.image.filename;
  }

  getStableId = (index) => `${index}`;

  getLayoutTypeForIndex = (index) => 1;

  setLayoutForType = (type, dim) => {
    const width1 = (screenWidth) / 3;
    dim.height = width1;
    dim.width = width1;
  };

  getGalleryDetails = params => {
    this.CameraRoll.getPhotos(params)
      .then(Response => {
        const lastCursor = Response.page_info.end_cursor;
        const hasNextpage = Response.page_info.has_next_page;

        const newRows = this.state.rows.concat(Response.edges);
        if (this.state.rows.length === 0) {
          this.setState({
            rows: newRows,
            lastCursor: lastCursor,
            hasNextpage: hasNextpage,
            finishedLoading: true,
            dataProvider: this._dataProvider.cloneWithRows(newRows)
          });
        } else {    
          this.setState({
            rows: newRows,
            dataProvider: this._dataProvider.cloneWithRows(newRows),
            lastCursor: lastCursor,
            hasNextpage: hasNextpage,
            finishedLoading: true
          });
        }
      })
      .catch(err => {
        this.setState({ finishedLoading: true })
        alert(err);
      });
  };

  onEndReached() {
    if (this.state.hasNextpage) {
      const params = {
        first: LOAD_IMAGES_AHEAD,
        assetType: "Photos",
        groupTypes: 'All',
        after: this.state.lastCursor
      };
      this.getGalleryDetails(params);
    }
  }
  
  onSelectedItems(item, index) {

    const selectedIndexes = this.state.selectedIndexes.includes(index) ?
      this.state.selectedIndexes.filter(data => data !== index) :
      this.state.selectedIndexes.concat([index]);

    const selectedFiles = this.state.selectedIndexes.includes(index) ?
      this.state.selectedFiles.filter(task => task.node.image.uri !== item.node.image.uri) :
      this.state.selectedFiles.concat([item])

    const { rows } = this.state;
    const mappedRows =  rows.map((r) => {
        return {
          ...r,
          isSelected: selectedIndexes
        }

    });

    this.setState(
      {
        selectedIndexes: selectedIndexes,
        selectedFiles: selectedFiles,
        dataProvider: this._dataProvider.cloneWithRows(mappedRows),
        rows: mappedRows
      }
    );
  }

  isSelected(index) {
    return this.state.selectedIndexes.includes(index);
  }

  getTimeDurationFromSec(secs) {
    var sec_num = parseInt(secs, 10);
    var hours = Math.floor(sec_num / 3600);
    var minutes = Math.floor(sec_num / 60) % 60;
    var seconds = sec_num % 60;

    return [hours, minutes, seconds]
      .map(v => (v < 10 ? "0" + v : v))
      .filter((v, i) => v !== "00" || i > 0)
      .join(":");
  }

  renderRow = (type, item, index) => {
    let seleIndex = this.state.selectedIndexes.findIndex(obj => obj === index)
    return <ImageCell
      disabled={this.state.selectedFiles.length == 10 && !this.isSelected(index)}
      selected={this.isSelected(index)}
      onSelect={() => this.onSelectedItems(item, index)}
      imageUrl={item.node.image.uri}
      selectedIndex={seleIndex}
      />
  };

  applyDimensions(layout) {
    const { width } = layout;
    this.setState({ width: width });
  }

  renderImageRow = (item, index) => {
    return (
      <TouchableOpacity
        style={{
          width: this.props.orientation == "portrait" ? 80 : 60,
          height: this.props.orientation == "portrait" ? 80 : 60,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Image
          source={{ uri: item.node.image.uri }}
          style={{ width: this.props.orientation == "portrait" ? 65 : 50, height: this.props.orientation == "portrait" ? 65 : 50}}
        />
      </TouchableOpacity>
    );
  };

  renderSettings() {
    return(
      <View style={styles.settingsContainer}>
            <Text style={{ fontWeight: "bold", color: "white", fontSize: 16, textAlign: "center" }}>
            {Strings.ios_select_more_photos}
            </Text>
            <TouchableOpacity
              style={[styles.settingsBtn, { height: 35, borderRadius: 17.5 }]}
              onPress={() => {
                openSettings()
              }}>
              <Text style={{ fontSize: 15, color: "white", alignSelf: "center", margin: 5, padding: 5 }}>Settings</Text>
            </TouchableOpacity>
          </View>
    )
  }
  
  render() {
    const {dataProvider,finishedLoading,permissionFlag} = this.state;
    const cellHeight = screenWidth / 3;
    
    const renderAhead = cellHeight * LOAD_IMAGES_AHEAD * 10;
    const onEndReachedThreshold = renderAhead;
    const settingFlag = dataProvider._data.length == 0 && finishedLoading && permissionFlag
    return (
        <View style={{ flex: 1, flexDirection: this.props.orientation == "portrait" ? "column":"row"}}>
        <View style={{ width: this.props.orientation == "portrait" ? "100%": "85%",height: (this.props.orientation == "portrait" && this.state.selectedFiles.length != 0) ? "85%":"100%"}}>
          <View
          onLayout={event => {
            this.applyDimensions(event.nativeEvent.layout);
          }}
          style={styles.container}
        >
          {!finishedLoading && (
            <ActivityIndicator
              size="large"
              color={Colors.primaryAccent}
              style={{ alignSef: "center" }}
            />
            )}
            {settingFlag && this.renderSettings()}
            { this.state.dataProvider._data.length !== 0 &&
              <RecyclerListView
                style={
                  {
                    flex: 1,
                    width: 500,
                    height: 800
                  }
                }
                key={this.props.orientation}
                onEndReached={this.onEndReached.bind(this)}
                keyExtractor={(item, index) => index.toString()}
                rowRenderer={this.renderRow}
                dataProvider={this.state.dataProvider}
                extraData={this.props.orientation}
                numColumns={this.props.rowItems}
                renderAheadOffset={renderAhead}
                onEndReachedThreshold={onEndReachedThreshold}
                disableRecycling={false}
                layoutProvider={this._layoutProvider}
              /> 
          
            }
        </View>
      </View>

      {this.state.selectedFiles.length > 0 ? (
        <View 
          style={{ 
              width:this.props.orientation == "portrait" ? "100%":"40%" ,
              flexDirection:  this.props.orientation == "portrait" ? "row":"column", 
              maxHeight: this.props.orientation == "portrait" ? 80 : "100%", 
              paddingLeft: this.props.orientation == "portrait" ? 0 : 20,
              justifyContent:"space-between",
          }}>
          {this.state.selectedFiles.length != 0 && 
            <View style={{ flex: 1}}>
              <FlatList
                horizontal={this.props.orientation == "portrait" ? true : false}
                data={this.state.selectedFiles}
                extraData={this.state.selectedFiles}
                renderItem={({ item, index }) => this.renderImageRow(item, index)}
                keyExtractor={item => item.id}
              />
            </View>
          }
          <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center'}} >
            <TouchableOpacity
              onPress={() => { this.props.onItemSelected(this.state.selectedFiles) }}
              style={styles.nextIcon}>
              <Image style={[{ height: 25, width: 25 }]} source={Images.next_video_icon} />
            </TouchableOpacity>
          </View>
        </View>) : null}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    height: "100%",
    width: "100%",
  },
  nextIcon:{
    height: 32,
    width: 32,
    margin: 10, 
    marginBottom:0,
    marginLeft:0,
    marginTop:35,
    marginRight:13,
    resizeMode: 'contain',
    alignSelf:"flex-end"
  },
   settingsBtn:{
    backgroundColor: Colors.primaryAccent,
    height:24,
     margin:10,
      minWidth: 50,
      borderRadius:12,
       justifyContent:"center",
       alignItems:"center"
      },
   settingsContainer:{
    width:"100%",
    height:80,
    alignSelf:"center",
    justifyContent:"center",
    flexDirection:"column",
    alignItems:"center",
    marginTop:200,
    padding:10
   }
});
