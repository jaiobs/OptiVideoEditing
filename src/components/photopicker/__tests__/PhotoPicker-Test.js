import React from 'react';
import renderer from 'react-test-renderer';
import PhotoPicker, { ImageCell } from '../PhotoPicker';
import {
  NativeModules,
} from "react-native";

const mockSelectedPhotos = jest.fn();

jest.mock('../PhotoPicker', () => {
  return jest.fn().mockImplementation(() => {
    return {onSelectedItems: mockSelectedPhotos};
  });
});

describe("PhotoPicker-Test",() => {
  beforeEach(() => {
    NativeModules.RNCCameraRoll = { getPhotos: (params) => {
      return Promise.resolve({
        page_info: {
          end_cursor: 0,
          has_next_page: 10
        },
        edges:[{
          node: {
            image: {
              url: "test-image.png"
            }
          }
        }]
      });
    } 
  } 
  });

  afterEach(() => {
    NativeModules.RNCameraManager = { converPHAssetIdentifierAsURLPath: (params) => {
      return Promise.resolve([ {
          node : {
              group_name : "All Photos",
               image : {
                  filename : "sideways_made_iphone11pro.mov",
                  height : 3482,
                  isStored : 1,
                  localPath : "file:///var/mobile/Media/DCIM/111APPLE/IMG_1740.MOV",
                  playableDuration : "2.771666666666667",
                  uri : "ph://A9C616F0-3000-429D-8E1D-A96DDC99DCA5/L0/001",
                  width : 2160,
              },
              location :{
              },
              timestamp : 1613877055,
              type : "video",
          }
        }
        ]);
    } 
  } 
  });

  it("should render the PhotoPicker", async () => {
      const tree = renderer
      .create(<PhotoPicker
        rowItems={5}
        orientation={'portrait'}
        />)
      .toJSON();
      
    expect(tree).toMatchSnapshot();
  })

  it("should render the initial ImageCell", () => {
    const tree = renderer
    .create(<ImageCell imageUrl={"test-image.png"} selected={false}/>)
    .toJSON();
    expect(tree).toMatchSnapshot();
})

  it("should render the selected ImageCell", () => {
    const tree = renderer
    .create(<ImageCell imageUrl={"test-image.png"} selected={true} selectedIndex={0}/>)
    expect(tree).toMatchSnapshot();
  })

 
  
})
