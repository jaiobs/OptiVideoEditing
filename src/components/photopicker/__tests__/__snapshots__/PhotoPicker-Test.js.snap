// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoPicker-Test should render the PhotoPicker 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#f9089b"
        hidesWhenStopped={true}
        size="large"
        style={
          Object {
            "alignSef": "center",
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`PhotoPicker-Test should render the initial ImageCell 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#f9089b",
      "borderWidth": 0,
      "height": "100%",
      "margin": 2,
      "opacity": 1,
      "padding": 1,
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "test-image.png",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "backgroundColor": "#f9089b",
        "borderRadius": 10,
        "height": 20,
        "justifyContent": "center",
        "position": "absolute",
        "right": 5,
        "top": 5,
        "width": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 12,
        }
      }
    >
      NaN
    </Text>
  </View>
</View>
`;

exports[`PhotoPicker-Test should render the selected ImageCell 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#f9089b",
      "borderWidth": 3,
      "height": "100%",
      "margin": 2,
      "opacity": 1,
      "padding": 1,
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "test-image.png",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "backgroundColor": "#f9089b",
        "borderRadius": 10,
        "height": 20,
        "justifyContent": "center",
        "position": "absolute",
        "right": 5,
        "top": 5,
        "width": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 12,
        }
      }
    >
      1
    </Text>
  </View>
</View>
`;
